usr.bin/lockf/lockf.1
.\"
.\" Copyright (C) 1998 John D. Polstra.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY JOHN D. POLSTRA AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL JOHN D. POLSTRA OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD: src/usr.bin/lockf/lockf.1,v 1.17 2005/07/31 03:30:46 keramida Exp $
.\"
.Dd July 7, 1998
.Os
.Dt LOCKF 1
.Sh NAME
.Nm lockf
.Nd execute a command while holding a file lock
.Sh SYNOPSIS
.Nm
.Op Fl ks
.Op Fl t Ar seconds
.Ar file
.Ar command
.Op Ar arguments
.Sh DESCRIPTION
The
.Nm
utility acquires an exclusive lock on a
.Ar file ,
creating it if necessary,
.Bf Em
and removing the file on exit unless explicitly told not to.
.Ef
While holding the lock, it executes a
.Ar command
with optional
.Ar arguments .
After the
.Ar command
completes,
.Nm
releases the lock, and removes the
.Ar file
unless the
.Fl k
option is specified.
.Bx Ns -style
locking is used, as described in
.Xr flock 2 ;
the mere existence of the
.Ar file
is not considered to constitute a lock.
.Pp
The following options are supported:
.Bl -tag -width Fl
.It Fl k
Causes the lock file to be kept (not removed) after the command
completes.
.It Fl s
Causes
.Nm
to operate silently.
Failure to acquire the lock is indicated only in the exit status.
.It Fl t Ar seconds
Specifies a timeout for waiting for the lock.
By default,
.Nm
waits indefinitely to acquire the lock.
If a timeout is specified with this option,
.Nm
will wait at most the given number of
.Ar seconds
before giving up.
A timeout of 0 may be given, in which case
.Nm
will fail unless it can acquire the lock immediately.
When a lock times out,
.Ar command
is
.Em not
executed.
.El
.Pp
In no event will
.Nm
break a lock that is held by another process.
.Sh EXIT STATUS
If
.Nm
successfully acquires the lock, it returns the exit status produced by
.Ar command .
Otherwise, it returns one of the exit codes defined in
.Xr sysexits 3 ,
as follows:
.Bl -tag -width F_CANTCREATX
.It Dv EX_TEMPFAIL
The specified lock file was already locked by another process.
.It Dv EX_CANTCREAT
The
.Nm
utility
was unable to create the lock file, e.g., because of insufficient access
privileges.
.It Dv EX_USAGE
There was an error on the
.Nm
command line.
.It Dv EX_OSERR
A system call (e.g., fork) failed unexpectedly.
.El
.Sh SEE ALSO
.Xr flock 2 ,
.Xr sysexits 3
.Sh HISTORY
A
.Nm
utility first appeared in
.Fx 2.2 .
.Sh AUTHORS
.An John Polstra Aq jdp@polstra.com .


usr.bin/lockf/lockf.c
/*
 * Copyright (C) 1997 John D. Polstra.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY JOHN D. POLSTRA AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL JOHN D. POLSTRA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD: src/usr.bin/lockf/lockf.c,v 1.16 2005/10/09 04:45:41 csjp Exp $");

#include <sys/types.h>
#include <sys/wait.h>

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <sysexits.h>
#include <unistd.h>

static void cleanup(void);
static void killed(int sig);
static void timeout(int sig);
static void usage(void);
static int wait_for_lock(const char *name, int flags);

static const char *lockname;
static int lockfd = -1;
static int keep;
static volatile sig_atomic_t timed_out;

/*
 * Execute an arbitrary command while holding a file lock.
 */
int
main(int argc, char **argv)
{
	int ch, silent, status, waitsec;
	pid_t child;

	silent = keep = 0;
	waitsec = -1;	/* Infinite. */
	while ((ch = getopt(argc, argv, "skt:")) != -1) {
		switch (ch) {
		case 'k':
			keep = 1;
			break;
		case 's':
			silent = 1;
			break;
		case 't':
		{
			char *endptr;
			waitsec = strtol(optarg, &endptr, 0);
			if (*optarg == '\0' || *endptr != '\0' || waitsec < 0)
				errx(EX_USAGE,
				    "invalid timeout \"%s\"", optarg);
		}
			break;
		default:
			usage();
		}
	}
	if (argc - optind < 2)
		usage();
	lockname = argv[optind++];
	argc -= optind;
	argv += optind;
	if (waitsec > 0) {		/* Set up a timeout. */
		struct sigaction act;

		act.sa_handler = timeout;
		sigemptyset(&act.sa_mask);
		act.sa_flags = 0;	/* Note that we do not set SA_RESTART. */
		sigaction(SIGALRM, &act, NULL);
		alarm(waitsec);
	}
	lockfd = wait_for_lock(lockname, O_NONBLOCK);
	while (lockfd == -1 && !timed_out && waitsec != 0)
		lockfd = wait_for_lock(lockname, 0);
	if (waitsec > 0)
		alarm(0);
	if (lockfd == -1) {		/* We failed to acquire the lock. */
		if (silent)
			exit(EX_TEMPFAIL);
		errx(EX_TEMPFAIL, "%s: already locked", lockname);
	}
	/* At this point, we own the lock. */
	if (atexit(cleanup) == -1)
		err(EX_OSERR, "atexit failed");
	if ((child = fork()) == -1)
		err(EX_OSERR, "cannot fork");
	if (child == 0) {	/* The child process. */
		close(lockfd);
		execvp(argv[0], argv);
		warn("%s", argv[0]);
		_exit(1);
	}
	/* This is the parent process. */
	signal(SIGINT, SIG_IGN);
	signal(SIGQUIT, SIG_IGN);
	signal(SIGTERM, killed);
	if (waitpid(child, &status, 0) == -1)
		err(EX_OSERR, "waitpid failed");
	return (WIFEXITED(status) ? WEXITSTATUS(status) : 1);
}

/*
 * Remove the lock file.
 */
static void
cleanup(void)
{

	if (keep)
		flock(lockfd, LOCK_UN);
	else
		unlink(lockname);
}

/*
 * Signal handler for SIGTERM.  Cleans up the lock file, then re-raises
 * the signal.
 */
static void
killed(int sig)
{

	cleanup();
	signal(sig, SIG_DFL);
	if (kill(getpid(), sig) == -1)
		err(EX_OSERR, "kill failed");
}

/*
 * Signal handler for SIGALRM.
 */
static void
timeout(int sig __unused)
{

	timed_out = 1;
}

static void
usage(void)
{

	fprintf(stderr,
	    "usage: lockf [-ks] [-t seconds] file command [arguments]\n");
	exit(EX_USAGE);
}

/*
 * Wait until it might be possible to acquire a lock on the given file.
 */
static int
wait_for_lock(const char *name, int flags)
{
	int fd;

	if ((fd = open(name, O_CREAT|O_RDONLY|O_EXLOCK|flags, 0666)) == -1) {
		if (errno == EINTR || errno == EAGAIN)
			return (-1);
		err(EX_CANTCREAT, "cannot open %s", name);
	}
	return (fd);
}
